docs/src/references/operation-workflow.md

```diff
@@ -0,0 +1,372 @@
+- A workflow implementation must not react on *no-op* states nor terminal states.
+  - The transition from a *no-op* state must be handled either by thin-edge as direct transition
+    or overridden by the user with domain-specific checks and actions.
+  - The terminal states, a.k.a `Successful` and `Failed`, are owned by the process which created the `Init` state (in practice, the mapper).
+    Only this process should clear the retained message state for an operation instance by sending an empty payload on command's topic.
+
+### Workflow Overriding
+
+Thin-edge provides a mechanism to override, extend and combine workflows.
+
+This mechanism is provided by the `tedge-agent` which gather a set of user-defined workflows
+and combined them with the builtin workflows implemented by the agent itself and the thin-edge operation plugins.
+
+Each workflow is defined using a TOML file that specifies:
+- the operation target and name defined as a topic filter
+  such as `te/device/main///cmd/software-update/+` or `te/device/main/service/+/cmd/restart/+`
+- the list of states
+- for each state:
+  - the state name as defined by the operation API
+  - the name of the state owner which is the process that is responsible for advancing a command at this state
+    (the default being `tedge`, meaning that thin-edge is responsible for running the state checks and actions)
+  - the set of states which can be an outcome for this state actions
+  - possible extra instructions on how to process the command at this stage, e.g.
+    - run a script
+    - emit an event
+    - raise an alarm
+  
+```toml title="file: workflow-example.toml"
+target = "te/device/main///cmd/configuration-update/+"
+
+[Init]
+  next = ["Download", "Failed"]
+  script = "/bin/schedule-configuration-update.sh"
+    
+[Download]
+  owner = tedge-configuration-plugin
+  next = ["Downloaded", "Failed"]
+
+[Downloaded]
+  script = "/bin/check-configuration-file.sh"
+  next = ["Install", "Failed"]
+
+[Install]
+  owner = tedge-configuration-plugin
+  next = ["Installed", "Failed"]
+        
+[Installed]
+  script = "/bin/check-configuration-update.sh"
+  next = ["Successful", "Failed"]
+
+[Successful]
+  event = "te/device/main///e/operations"
+  next = []
+  
+[Failed]
+  alarm = "te/device/main///a/operations"
+  next = []
+```
+
+Thin-edge combines all these workflows to determine what has to be done
+when a state message is published for a command on one topic matching the global topic filter for commands,
+i.e. `te/+/+/+/+/cmd/+/+`. Given an actual command state message, thin-edge 
+- keeps only the workflows which target topic filter matches the topic of the message
+- then keeps only the workflows that define a behavior for the state of the message payload
+- and then uses priority rules to select a single workflow: the workflow ruling this command at that stage
+
+The priority rules give a higher priority to the workflow that are user-defined than to those pre-defined by thin-edge.
```

Comment:
I like the concept of supporting priorities, to allow overriding workflows, and it is done in a very "linux friendly" format.